chain/chain/src/flat_storage_resharder.rs

```diff
@@ -0,0 +1,1008 @@
+                parent_shard,
+                FlatStorageStatus::Ready(FlatStorageReadyStatus { flat_head }),
+            );
+            // Remove children shards leftovers.
+            for child_shard in [left_child_shard, right_child_shard] {
+                store_update.remove_flat_storage(child_shard);
+            }
+        }
+    }
+    store_update.commit().unwrap();
+    // Terminate the resharding event.
+    *resharder.resharding_event.lock().unwrap() = None;
+}
+
+/// Copies a key-value pair to the correct child shard by matching the account-id to the provided shard layout.
+fn copy_kv_to_child(
+    status: &SplittingParentStatus,
+    key: Vec<u8>,
+    value: FlatStateValue,
+    store_update: &mut FlatStoreUpdateAdapter,
+    account_id_parser: impl FnOnce(&[u8]) -> Result<AccountId, std::io::Error>,
+) -> Result<(), std::io::Error> {
+    let SplittingParentStatus { left_child_shard, right_child_shard, shard_layout, .. } = &status;
+    // Derive the shard uid for this account in the new shard layout.
+    let account_id = account_id_parser(&key)?;
+    let new_shard_id = account_id_to_shard_id(&account_id, shard_layout);
+    let new_shard_uid = ShardUId::from_shard_id_and_layout(new_shard_id, &shard_layout);
+
+    // Sanity check we are truly writing to one of the expected children shards.
+    if new_shard_uid != *left_child_shard && new_shard_uid != *right_child_shard {
+        let err_msg = "account id doesn't map to any child shard!";
+        error!(target: "resharding", ?new_shard_uid, ?left_child_shard, ?right_child_shard, ?shard_layout, ?account_id, err_msg);
+        return Err(std::io::Error::new(std::io::ErrorKind::InvalidData, err_msg));
+    }
+    // Add the new flat store entry.
+    store_update.set(new_shard_uid, key, Some(value));
+    Ok(())
+}
+
+/// Struct to describe, perform and track progress of a flat storage resharding.
+#[derive(Clone, Debug)]
+pub enum FlatStorageReshardingEventStatus {
+    /// Split a shard.
+    /// Includes the parent shard uid and the operation' status.
+    SplitShard(ShardUId, SplittingParentStatus),
+}
+
+/// Status of a flat storage resharding task.
+#[derive(Clone, Debug, Copy, Eq, PartialEq)]
+pub enum FlatStorageReshardingTaskStatus {
+    Successful,
+    Failed,
+    Interrupted,
+}
+
+/// Helps control the flat storage resharder operation. More specifically,
+/// it has a way to know when the background task is done or to interrupt it.
+#[derive(Clone)]
+pub struct FlatStorageResharderController {
+    /// Resharding handle to control interruption.
+    handle: ReshardingHandle,
+    /// This object will be used to signal when the background task is completed.
+    /// A value of `true` means that the operation completed successfully.
+    completion_sender: Sender<FlatStorageReshardingTaskStatus>,
+    /// Corresponding receiver for `completion_sender`.
+    pub completion_receiver: Receiver<FlatStorageReshardingTaskStatus>,
```

Comment:
Umm, why exactly would the `FlatStorageResharderController` hold the completion_receiver? Shouldn't that be sent to someone who is waiting for the call the resharding of flat storage to be finished? Someone in client?

Also, is this the patter we would like to follow? I thought the convention was to send actix messages once jobs were done? Btw, where is the job of flat storage resharding being done? As in, which actor is it a part of?

Reply:
Would we reload memtrie immediately after FS is ready or delay until node restart?

Reply:
The current understanding is to do memtrie reload FS is ready. Not set in stone though